File: 1277. Count Square Submatrices with All Ones/dp.py
class Solution:
    def countSquares(self, grid: List[List[int]]) -> int:
        H, W = len(grid), len(grid[0])
        dp = [[0 for x in range(W)] for y in range(H)];
        res = 0
        
        for y in range(H):
            for x in range(W):
                if grid[y][x] == 1:
                    if(y == 0 or x == 0):
                        dp[y][x] = 1
                    else:
                        dp[y][x] = min(
                            dp[y-1][x],
                            dp[y][x-1],
                            dp[y-1][x-1]
                        ) + 1
                        
                    res += dp[y][x]
        
        return res;
        

File: 0063. Unique Paths II/dp.java
class Solution {
    public int uniquePathsWithObstacles(int[][] grid) {
        
        int H = grid.length, W = grid[0].length;
        
        int[][] dp = new int[H][W];
        
        for(int y = 0; y < H; y++){
            for(int x = 0; x < W; x++){
                if(y == 0 && x == 0 && grid[y][x] != 1){
                    dp[0][0] = 1;
                    continue;
                }
                dp[y][x] = 0;
                if(grid[y][x] == 1) continue;
                
                if(y != 0) dp[y][x] += dp[y-1][x];
                if(x != 0) dp[y][x] += dp[y][x-1];
            }
        }
        
        return dp[H-1][W-1];
    }
}